Replace UMapCompact with UMap and add tests

aniketd opened the pull request:
# Description

Closes #3272 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the `CHANGELOG.md` for affected package
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

aniketd commented:
2 tests are failing still, unexpectedly. I'm looking into it :+1:

JaredCorduan reviewed (approved):
love it! thank you!

aniketd commented:
Those tests should now pass :crossed_fingers:

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -363,6 +364,10 @@ compactRewView x = Map.map rdReward $ unUnify (RewardDeposits x)
 
+-- | Materialize the RDPairs Map from a 'UMap'
+rdPairView :: UMap c -> Map.Map (Credential 'Staking c) RDPair
+rdPairView x = unUnify (RewardDeposits x)
```
Could you deprecate `rewards` in favor of this function. `rdPairView` is a better name anyways.

aniketd replied:
Taking a closer look, I realized `rewards` and `rdPairView` are slightly different. Do you mean `rewView` and `depositView`?

aniketd replied:
After thinking even further, do you mean I could rename `rewards` to `rdPairs` to reflect the function semantics?

lehins replied:
I also didn't realize until now that they are slightly different. However the name `rewards` was invalid, so renaming the function was a good call. Thank you for doing it.

lehins commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/UMapSpec.hs:
My guess is you didn't do `git mv`? Cause the git history seems to have disappeared.

Can you please fix it, cause it will make the review easier and will help with history in general

aniketd replied:
I did `git mv` but then squashed all into one. I will split it out and retain an individual commit with just that. Thanks!

lehins commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/UMapSpec.hs:
```diff
@@ -0,0 +1,400 @@
+   in
+    (Map.size rdPairs == rdPairsSize)
+      && (Map.size delegs == delegsSize)
+      && (Map.size ptrs == ptrsSize)
```
Let's improve failure reporting a bit
```suggestion
  Expectation
sizeTest (rdPairs, delegs, ptrs) = do
  let
    umap = unify rdPairs delegs ptrs
    rdPairsSize = size (RewardDeposits umap)
    delegsSize = size (Delegations umap)
    ptrsSize = size (Ptrs umap)
  Map.size rdPairs `shouldBe` rdPairsSize
  Map.size delegs `shouldBe` delegsSize
  Map.size ptrs `shouldBe` ptrsSize
```

lehins commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/UMapSpec.hs:
```diff
@@ -0,0 +1,400 @@
+    delegs' = delView umap
+    ptrs' = ptrView umap
+   in
+    rdPairs == rdPairs' && delegs == delegs' && ptrs == ptrs'
```
Same here:
```suggestion
  Expectation
unifyRoundTripTo (rdPairs, delegs, ptrs) = do
  let umap = unify rdPairs delegs ptrs
  rdPairView umap `shouldBe` rdPairs
  delView umap `shouldB` delegs
  ptrView umap `shouldBe` ptrs
```

lehins commented on libs/cardano-ledger-core/test/Test/Cardano/Ledger/UMapSpec.hs:
```diff
@@ -0,0 +1,400 @@
+spec :: Spec
+spec = do
+  describe "UMap" $ do
+    context "Invariant" $ do
```
Why `context` in one place and `describe` in another? I think using the same `describe` for consistency everywhere is a good practice.

aniketd replied:
I thought that's how `context` was supposed to be used, as a sub-grouping of a main test that is `describe`d. But you are right, other than being a bit more readable it hardly adds any value. But it brings more potential confusion as `context` isn't that popular I guess. I will change it to just using `describe`. :+1:

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
   arbitrary = UMap <$> arbitrary <*> arbitrary
   shrink = genericShrink
```
Could you remove this one. Generic should not be used with `Map` (consequently with `UMap` either)
```suggestion
```

aniketd replied:
Noted!

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+    )
+genValidTriples = resize 50 $ do
+  nCreds <- chooseInt (1, 50)
+  creds :: [Credential 'Staking StandardCrypto] <- vectorOf nCreds arbitrary
```
Empty should not be excluded
```suggestion
  creds :: [Credential 'Staking StandardCrypto] <-  arbitrary
  let nCreds = length creds
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+  creds :: [Credential 'Staking StandardCrypto] <- vectorOf nCreds arbitrary
+  rdPairs :: [RDPair] <- vectorOf nCreds arbitrary
+  delegs :: [KeyHash 'StakePool StandardCrypto] <- vectorOf nCreds arbitrary
+  ptrs :: [Ptr] <- vectorOf nCreds arbitrary
```
Pointer map does not have to have the same number of elements as the credentials
```suggestion
  ptrs :: [Ptr] <- arbitrary
```

aniketd replied:
Understood! The call to `elements` however, throws an exception that its argument can't be empty. This ends up agreeing with our intention to generate non-empty `UMap`s only, as the tests that use `genInvariant` indeed expect a non-empty `UMap`. So I have renamed it to `genInvariantNonEmpty`.

```haskell
genValidTriplesNonEmpty = scale (* 2) $ do
  Positive nCreds <- arbitrary
  nPtrs <- chooseInt (1, nCreds)
  creds :: [Credential 'Staking StandardCrypto] <- vectorOf nCreds arbitrary
  rdPairs :: [RDPair] <- vectorOf nCreds arbitrary
  delegs :: [KeyHash 'StakePool StandardCrypto] <- vectorOf nCreds arbitrary
  ptrs :: [Ptr] <- vectorOf nPtrs arbitrary
  pure
    ( Map.fromList $ zip creds rdPairs
    , Map.fromList $ zip creds delegs
    , Map.fromList $ zip ptrs creds
    )
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+    , Map.Map (Credential 'Staking StandardCrypto) (KeyHash 'StakePool StandardCrypto)
+    , Map.Map Ptr (Credential 'Staking StandardCrypto)
+    )
+genValidTriples = resize 50 $ do
```
Almost always it is better to `scale`, rather than `resize`, because if someone does `scale (*10) $ genValidTriples`, they will not get the size `50`, but the desired `500` (or rather `600` when `(* 2)` is used)
```suggestion
genValidTriples = scale (* 2) $ do
```

aniketd replied:
Thanks for correcting me here again. :slightly_smiling_face:

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+  a <- arbitrary
+  if a `elem` as
+    then genExcluding as
+    else pure a
```
Avoid `elem` on lists because it is an O(n) operation, either use a `Set` or a `Map`. Looks like you are already using a Map, so:
```suggestion
genExcludingKey :: (Ord k, Arbitrary k) => Map k a -> Gen k
genExcludingKey as = do
  a <- arbitrary
  if a `Map.member` as
    then genExcludingKey as
    else pure a
```

aniketd replied:
Awesome! Thanks! :slightly_smiling_face:

lehins commented on libs/cardano-ledger-core/CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## 1.1.0.0
 
+* Rename `UMapCompact` to `UMap` and add tests for it.
```
`1.1.0.0` has now been released, so this PR should start a `1.2.0.0`. This will require a bump in cabal file as well.

You should record that the module name has changed. The way it is written now looks like a type rename. 
```suggestion
* Rename `Cardano.Ledger.UMapCompact` to `Cardano.Ledger.UMap` and add tests for it.
```
Also there is no mention of new function: `rdPairView`

Moreover, every package that has been affected by this rename, must have its lower bound be updated.

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+  cred <-
+    oneof
+      [ elements creds
+      , genExcluding creds
```
```suggestion
      , genExcludingKey tripmap
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Arbitrary.hs:
```diff
@@ -467,6 +472,62 @@ instance Crypto c => Arbitrary (UMap c) where
+  ptr <-
+    oneof
+      [ elements ptrs
+      , genExcluding ptrs
```
```suggestion
      , genExcludingKey ptrmap
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
It would be nice to reduce breakage and promote easier migration by adding a module in a separate commit `Cardano.Ledger.UMapCompact` that re-exports `Cardano.Ledger.UMap`, but is marked as deprecated.

lehins commented on libs/cardano-data/cardano-data.cabal:
```diff
@@ -20,7 +20,6 @@ library
         Data.CanonicalMaps
         Data.Pulse
         Data.MapExtras
-        Data.UMap
```
Module has been removed. That is a breaking change, therefore it needs to be reflected in a cabal version and the changelog. 

Moreover, this removal affects older versions of `set-algebra`, so if you feel adventurous you could [add a revision](https://github.com/input-output-hk/cardano-haskell-packages#how-to-add-a-new-package-metadata-revision) on the already released [version of `set-algebra` ](https://github.com/input-output-hk/cardano-haskell-packages/blob/main/_sources/set-algebra/1.0.0.0/meta.toml) and add an upper bound on the `cardano-data`, but only after this PR is merged, of course.

lehins commented on libs/set-algebra/src/Control/Iterate/BaseTypes.hs:
```diff
@@ -278,41 +270,6 @@ instance Iter Map.Map where
-
-instance
-  (Ord coin, Ord cred, Ord ptr) =>
-  Iter (UM.View coin cred pool ptr)
```
Removing instances is a breaking change. Must be reflected in the changelog and in a version bump.

lehins reviewed (changes requested):
Looking great. Some issues with versions and bounds, but this is a totally new practice for everyone on the team, so it will take a while to get used to.

aniketd committed:
Remove 'Data.UMap'.

- Also, remove use of it from set-algebra. Most of it were unused instances.
- Bump versions of both `cardano-data` and `set-algebra` and update their changelogs.

This is in preparation for renaming `Cardano.Ledger.UMapCompact` back to `UMap` in
subsequent commits.

aniketd committed:
Rename `Cardano.Ledger.UMapComact` to `Cardano.Ledger.UMap`.

Also, bump `cardano-ledger-core` version to 1.2.0.0

aniketd committed:
Rename `UMapCompactSpec` to `UMapSpec`.

These tests are not likely to compile without errors yet, so just
the file is moved, but not added to the cabal file.

aniketd committed:
Update UMapSpec tests and add them to cabal.

Add missing required arbitrary instances.

aniketd committed:
Add UMapComact as deprecated. Update changelog for entire PR.

The versions don't need bumping as we have already done that in a
previous commit.

aniketd commented:
@lehins This is ready for another review!

lehins commented on libs/cardano-ledger-core/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+- Deprecate `Cardano.Ledger.UMapCompact` in favor of `Cardano.Ledger.UMap` and add tests for it. #3371
+  - Add `Cardano.Ledger.UMap.rdPairView` to view the reward-deposits pair from the `UMap`.
+  - Rename `Cardano.Ledger.UMap.rewards` to `rdPairs`.
```
`rewards` function was never exported, so this change should not be added to the changelog
```suggestion
```

aniketd replied:
Oh! Got it! Thanks!

lehins reviewed (approved):
Awesome work. Thank you.

lehins commented:
@aniketd Looks like there is a test failure

aniketd committed:
Fix bug in UMap.insertWith' and generators for insert-delete roundtrip

lehins reviewed (approved):
:+1:
